fix(agent): handle None response from retry exhaustion to prevent crashes

gajeshbhat opened the pull request:
## Description

Fixes issue #438 where Testflinger agents show EXITED status in **supervisorctl** and cannot recover themselves.

### Problem

When the Testflinger server experiences issues (like repeated 503 errors), the agent's HTTP client retry mechanism can return `None` instead of a proper `Response` object after exhausting all retries. The existing error handling code checked `if not job_request:` but then tried to access `job_request.status_code`, causing an `AttributeError: 'NoneType' object has no attribute 'status_code'`.

This unhandled exception caused the agent process to crash, leading to:
- Agents showing EXITED status in supervisorctl
- Manual intervention required to restart agents
- Loss of job processing capability

### Root Cause

The issue stems from the ambiguity in Boolean evaluation of `requests.Response` objects:
- `None` evaluates to `False` (retry exhaustion)  
- `Response` with status >= 400 also evaluates to `False` (HTTP error)

Both cases triggered the same error handling path, but only the Response case has a `status_code` attribute.

### Boolean Evaluation Issue

The `requests.Response` object implements a custom `__bool__` method that returns `False` for HTTP status codes >= 400 ([requests documentation](https://requests.readthedocs.io/en/latest/api/#requests.Response)). This creates an ambiguous condition where both cases evaluate as falsy:

```python
# Both evaluate to True in: if not job_request:
job_request = None                    # Retry exhaustion → None
job_request = Response(status=503)    # HTTP error → False due to __bool__
```

### Requests Retry Behaviour

When using `urllib3.util.retry.Retry` with `requests.Session`, the retry mechanism can return `None` in edge cases after exhausting all retry attempts, particularly with network-level failures ([urllib3 retry documentation](https://urllib3.readthedocs.io/en/stable/reference/urllib3.util.html#urllib3.util.retry.Retry)). The original code assumed all HTTP operations would return a `Response` object, but this assumption breaks during severe connectivity issues.

The fix described in the section below explicitly distinguishes between these two falsy states:
```python
if not job_request:
    if job_request is None:  # Retry exhaustion
        logger.error("No response received")
    else:  # HTTP error response
        logger.error("HTTP %d", job_request.status_code)
```

### Solution

Updated error handling in 5 client methods to explicitly check for `None` before accessing `status_code`:

- `post_status_update()` - The main method causing reported crashes
- `post_result()`
- `get_result()` 
- `repost_job()`
- `post_artifacts()`

## Resolved issues
Fixes #438 (Resolves #CERTTF-474 Internally)

## Documentation
Bug fix. No documentation update required

## Web service API changes
None.  Added better error handling in a bugfix

## Tests

- Added 4 new unit tests to verify None response handling
- All existing tests continue to pass (82/82 agent tests)
- Verified with other components (CLI: 67/67, Server: 116/116, Common: 3/3)
- All linting checks pass

gajeshbhat committed:
fix(agent): handle None response from retry exhaustion to prevent crashes

Fix AttributeError when session.post() returns None after retry exhaustion.
When the server returns repeated 503 errors, the requests retry mechanism
can return None instead of a Response object, causing agents to crash when
accessing status_code attribute.

Fixed methods:
- post_status_update()
- post_result()
- get_result()
- repost_job()
- post_artifacts()

Resolves agents showing EXITED status in supervisorctl.

Fixes #438 (Fixes #CERTTF-474 Internally)

Fix AttributeError when session.post() returns None after retry exhaustion.
When the server returns repeated 503 errors, the requests retry mechanism
can return None instead of a Response object, causing agents to crash when
accessing status_code attribute.

Fixed methods:
- post_status_update()
- post_result()
- get_result()
- save_artifacts()

Resolves agents showing EXITED status in supervisorctl.
Fixes #438 (Fixes #CERTTF-474 Internally)

gajeshbhat commented:
CI failed because of the policy check. Need maintainers' approval to rerun.

gajeshbhat commented:
@pedro-avalos @rene-oromtz Can you kindly look at this PR and let me know if my submission will be accepted?

gajeshbhat commented:
**Run charm test** CI check hangs and did not run successfully. The tests I added only cover the None cases (when job_request is None), but the else branches (when job_request is a Response object with status >= 400) are NOT covered yet. I will add tests to enchance coverage.

ajzobro commented on agent/src/testflinger_agent/client.py:
```diff
@@ -215,11 +223,18 @@ def get_result(self, job_id):
+                    "No response received",
+                )
+            else:
+                logger.error(
```
And here.

ajzobro reviewed (changes requested):
Thank you for your contribution and your patience, we are sorry it has taken so long to get around to reviewing it. 

Please add tests for:

post_result() with a response object that has a status_code (lines 202-207)
get_result() with a response object that has a status_code (lines 233-237)
save_artifacts() with a response object that has a status_code (lines 335-340)

gajeshbhat commented:
> Thank you for your contribution and your patience, we are sorry it has taken so long to get around to reviewing it.
> 
> Please add tests for:
> 
> post_result() with a response object that has a status_code (lines 202-207) get_result() with a response object that has a status_code (lines 233-237) save_artifacts() with a response object that has a status_code (lines 335-340)

Thank you for taking a look. I made some changes and pushed them. Let me know if you have any questions.

gajeshbhat committed:
test(agent): add status_code branch tests and use @patch decorators

Signed-off-by: Gajesh Bhat <gajeshbht@gmail.com>

gajeshbhat commented:
@ajzobro I have fixed the conflicts and rebased. Can you take a look at my changes based on your last comments ?

gajeshbhat commented:
> For tracking, this corresponds with CERTTF-474 in the Jira plan.

Thank you @ajzobro Anything else we need to update before let the CI Pipelines run ? I see they are failing as they were not approved before run.

ajzobro commented:
If you rebase or merge in the latest `main` changes the unit tests should run now.  We've updated the action so that they will run on the GH servers.  Can you test that out.

gajeshbhat commented:
> If you rebase or merge in the latest `main` changes the unit tests should run now. We've updated the action so that they will run on the GH servers. Can you test that out.

Thank you for the update @ajzobro . I rebased with upstream/main and repushed my changes. All tests pass except Run charm integration tests job which still fails because of a similar error.

```
A job started hook has been configured by the self-hosted runner administrator
Run '/home/ubuntu/actions-runner/pre-job.sh'
<13>May  1 07:19:11 ubuntu: IS_INTERNAL_PR: false (head: gajeshbhat/testflinger, base: canonical/testflinger)
<13>May  1 07:19:11 ubuntu: PR HEAD repo: gajeshbhat/testflinger, PR BASE repo: canonical/testflinger
<13>May  1 07:19:11 ubuntu: Fork PR or missing repository detected - performing contributor check
<13>May  1 07:19:11 ubuntu: Author association: CONTRIBUTOR, is allowed: false
<11>May  1 07:19:11 ubuntu: Insufficient user authorization - only OWNER, MEMBER, or COLLABORATOR may run workflows
<13>May  1 07:19:11 ubuntu: Contributor check failed - aborting job execution
Error: Process completed with exit code 1.
```

ajzobro commented:
Upon further review, the original issue was fixed by commit https://github.com/canonical/testflinger/commit/e23bd576aa5b3b319777b70adc3f5dd7df244393 and the `self.session.post` cannot return None (and was not returning None when the bug was discovered) so the proposed fix turns out to no longer be needed.
